### PR TITLE
llama3 custom regex split

### PR DESCRIPTION
**Pull request by jaime-m-p**
Implementation of `unicode_regex_split_custom_llama3()`.

### CODE REVIEW

**Comment by ggerganov** [comment]
I had to disable the custom gpt-2 implementation because it turns out there is a bug:

https://github.com/ggerganov/llama.cpp/blob/120cf37d546a82042bd1ad5894e4814a0efa8086/unicode.cpp#L430-L434

Seems like your custom llama-3 implementation is correct, because the tests are passing. Good job 👍 

Let's merge this after we merge #6920 later today

**Comment by slaren** [comment]
It may be worth looking into header only regex libraries that could be easily incorporated into llama.cpp. For example, https://github.com/hanickadot/compile-time-regular-expressions may be an option. `std::regex` apparently is notoriously bad.

**Comment by reneleonhardt** [comment]
I wouldn't rule out a real battle-tested regex engine like https://github.com/kkos/oniguruma#usage.
You import one header file and that's it, isn't it? 😅
hanickadot seems to require at least C++17, llama.cpp enforces C++11.

**Comment by slaren** [comment]
@reneleonhardt I don't think this library can be built as a single header (or even as a single header + single source file). `ctre` looks particularly interesting because it can be bundled as a single header file without additional dependencies.

**Comment by bartowski1182** [comment]
Not obvious from the code or PR, does this affect conversion, inference, or both?

File: unicode.cpp
Comment (ggerganov) [inline unicode.cpp:535]:
```suggestion
    } else if (
            regex_expr == "(?i:'s|'t|'re|'ve|'m|'ll|'d)|[^\\r\\n\\p{L}\\p{N}]?\\p{L}+|\\p{N}{1,3}| ?[^\\s\\p{L}\\p{N}]+[\\r\\n]*|\\s*[\\r\\n]+|\\s+(?!\\S)|\\s+" ||
            regex_expr == "(?:'[sS]|'[tT]|'[rR][eE]|'[vV][eE]|'[mM]|'[lL][lL]|'[dD])|[^\\r\\n\\p{L}\\p{N}]?\\p{L}+|\\p{N}{1,3}| ?[^\\s\\p{L}\\p{N}]+[\\r\\n]*|\\s*[\\r\\n]+|\\s+(?!\\S)|\\s+") {

```

**Comment by ggerganov** [comment]
Can you suggest a large text corpus that is as diverse as possible? I want to verify that the tokenization matches 100% the `transformers` one. I'm testing with `wiki.train` and it matches, but I feel it is not diverse enough.

File: unicode.cpp
Comment (ggerganov) [inline unicode.cpp:227]: perf note: this call can be avoided since we already computed the codepoints in `unicode_regex_split()`. We just need to pass a const reference in the custom functions

**Comment by ggerganov** [comment]
Please apply the following patch to match the style better:

<details>

```diff
diff --git a/unicode.cpp b/unicode.cpp
index 9abf83d3..7f37d5eb 100644
--- a/unicode.cpp
+++ b/unicode.cpp
@@ -228,7 +228,6 @@ static std::vector<size_t> unicode_regex_split_custom_gpt2(const std::string & t
 
     size_t start = 0;
     for (auto offset : offsets) {
-
         const size_t offset_ini = start;
         const size_t offset_end = start + offset;
         assert(offset_end <= cpts.size());
@@ -246,8 +245,9 @@ static std::vector<size_t> unicode_regex_split_custom_gpt2(const std::string & t
         auto _add_token = [&] (const size_t end) -> size_t {
             assert(_prev_end <= end && end <= offset_end);
             size_t len = end - _prev_end;
-            if(len > 0)
+            if (len > 0) {
                 bpe_offsets.push_back(len);
+            }
             _prev_end = end;
             //if(len) {
             //    std::string s = "";
@@ -258,7 +258,7 @@ static std::vector<size_t> unicode_regex_split_custom_gpt2(const std::string & t
             return len;
         };
 
-        for(size_t pos = offset_ini; pos < offset_end; /*pos++*/ ) {
+        for (size_t pos = offset_ini; pos < offset_end; /*pos++*/ ) {
             const char32_t cpt = _get_cpt(pos);
             const int cpt_type = _get_cpt_type(pos);
 
@@ -268,7 +268,8 @@ static std::vector<size_t> unicode_regex_split_custom_gpt2(const std::string & t
                 if (cpt_next == 's' || cpt_next == 't' || cpt_next == 'm' || cpt_next == 'd') {
                     pos += _add_token(pos+2);
                     continue;
-                } else if (pos+2 < offset_end) {
+                }
+                if (pos+2 < offset_end) {
                     char32_t cpt_next_next = _get_cpt(pos+2);
                     if ((cpt_next == 'r' && cpt_next_next == 'e') ||
                         (cpt_next == 'v' && cpt_next_next == 'e') ||
@@ -283,24 +284,27 @@ static std::vector<size_t> unicode_regex_split_custom_gpt2(const std::string & t
             // regex: <space>?\p{L}+
             if (cpt2_type == CODEPOINT_TYPE_LETTER) {
                 pos += (cpt == ' ');
-                while(cpt2_type == CODEPOINT_TYPE_LETTER)
+                while (cpt2_type == CODEPOINT_TYPE_LETTER) {
                     cpt2_type = _get_cpt_type(++pos);
+                }
                 _add_token(pos);
                 continue;
             }
             // regex: <space>?\p{N}+
             if (cpt2_type == CODEPOINT_TYPE_DIGIT) {
                 pos += (cpt == ' ');
-                while(cpt2_type == CODEPOINT_TYPE_DIGIT)
+                while (cpt2_type == CODEPOINT_TYPE_DIGIT) {
                     cpt2_type = _get_cpt_type(++pos);
+                }
                 _add_token(pos);
                 continue;
             }
             // regex: <space>?[^\s\p{L}\p{N}]+
             if (cpt2_type != CODEPOINT_TYPE_WHITESPACE && cpt2_type != CODEPOINT_TYPE_LETTER && cpt2_type != CODEPOINT_TYPE_DIGIT && cpt2_type != CODEPOINT_TYPE_UNIDENTIFIED) {
                 pos += (cpt == ' ');
-                while(cpt2_type != CODEPOINT_TYPE_WHITESPACE && cpt2_type != CODEPOINT_TYPE_LETTER && cpt2_type != CODEPOINT_TYPE_DIGIT && cpt2_type != CODEPOINT_TYPE_UNIDENTIFIED)
+                while (cpt2_type != CODEPOINT_TYPE_WHITESPACE && cpt2_type != CODEPOINT_TYPE_LETTER && cpt2_type != CODEPOINT_TYPE_DIGIT && cpt2_type != CODEPOINT_TYPE_UNIDENTIFIED) {
                     cpt2_type = _get_cpt_type(++pos);
+                }
                 _add_token(pos);
                 continue;
             }
@@ -308,17 +312,17 @@ static std::vector<size_t> unicode_regex_split_custom_gpt2(const std::string & t
             size_t num_whitespaces = 0;
             while (_get_cpt_type(pos+num_whitespaces) == CODEPOINT_TYPE_WHITESPACE) {
                 num_whitespaces++;
-                }
+            }
 
             // regex: \s+(?!\S)
-            if(num_whitespaces > 1 && _get_cpt(pos+num_whitespaces) != 0) {
+            if (num_whitespaces > 1 && _get_cpt(pos+num_whitespaces) != 0) {
                 pos += num_whitespaces - 1;
                 _add_token(pos);
                 continue;
             }
 
             // regex: \s+
-            if(num_whitespaces > 0) {
+            if (num_whitespaces > 0) {
                 pos += num_whitespaces;
                 _add_token(pos);
                 continue;
@@ -341,7 +345,6 @@ static std::vector<size_t> unicode_regex_split_custom_llama3(const std::string &
 
     size_t start = 0;
     for (auto offset : offsets) {
-
         const size_t offset_ini = start;
         const size_t offset_end = start + offset;
         assert(offset_end <= cpts.size());
@@ -359,8 +362,9 @@ static std::vector<size_t> unicode_regex_split_custom_llama3(const std::string &
         auto _add_token = [&] (const size_t end) -> size_t {
             assert(_prev_end <= end && end <= offset_end);
             size_t len = end - _prev_end;
-            if(len > 0)
+            if (len > 0) {
                 bpe_offsets.push_back(len);
+            }
             _prev_end = end;
             //if(len) {
             //    std::string s = "";
@@ -371,7 +375,7 @@ static std::vector<size_t> unicode_regex_split_custom_llama3(const std::string &
             return len;
         };
 
-        for(size_t pos = offset_ini; pos < offset_end; /*pos++*/ ) {
+        for (size_t pos = offset_ini; pos < offset_end; /*pos++*/ ) {
             const char32_t cpt = _get_cpt(pos);
             const int cpt_type = _get_cpt_type(pos);
 
@@ -381,7 +385,8 @@ static std::vector<size_t> unicode_regex_split_custom_llama3(const std::string &
                 if (cpt_next == 's' || cpt_next == 't' || cpt_next == 'm' || cpt_next == 'd') {
                     pos += _add_token(pos+2);
                     continue;
-                } else if (pos+2 < offset_end) {
+                }
+                if (pos+2 < offset_end) {
                     char32_t cpt_next_next = unicode_tolower(_get_cpt(pos+2));
                     if ((cpt_next == 'r' && cpt_next_next == 'e') ||
                         (cpt_next == 'v' && cpt_next_next == 'e') ||
@@ -394,10 +399,11 @@ static std::vector<size_t> unicode_regex_split_custom_llama3(const std::string &
 
             // regex: [^\r\n\p{L}\p{N}]?\p{L}+  //####FIXME: the first \p{L} is correct?
             if (cpt != '\r' && cpt != '\n' && /*cpt_type != CODEPOINT_TYPE_LETTER &&*/ cpt_type != CODEPOINT_TYPE_DIGIT) {
-                if(cpt_type == CODEPOINT_TYPE_LETTER || _get_cpt_type(pos+1) == CODEPOINT_TYPE_LETTER) {  // one or more letters
+                if (cpt_type == CODEPOINT_TYPE_LETTER || _get_cpt_type(pos+1) == CODEPOINT_TYPE_LETTER) {  // one or more letters
                     pos++;
-                    while(_get_cpt_type(pos) == CODEPOINT_TYPE_LETTER)
+                    while (_get_cpt_type(pos) == CODEPOINT_TYPE_LETTER) {
                         pos++;
+                    }
                     _add_token(pos);
                     continue;
                 }
@@ -406,7 +412,7 @@ static std::vector<size_t> unicode_regex_split_custom_llama3(const std::string &
             // regex: \p{N}{1,3}
             if (cpt_type == CODEPOINT_TYPE_DIGIT) {
                 size_t ini = pos;
-                while(_get_cpt_type(pos) == CODEPOINT_TYPE_DIGIT) {
+                while (_get_cpt_type(pos) == CODEPOINT_TYPE_DIGIT) {
                     if (++pos - ini >= 3 ) {
                         _add_token(pos);
                         ini = pos;
@@ -420,11 +426,13 @@ static std::vector<size_t> unicode_regex_split_custom_llama3(const std::string &
             int cpt2_type = (cpt == ' ' ? _get_cpt_type(pos+1) : cpt_type);
             if (cpt2_type != CODEPOINT_TYPE_WHITESPACE && cpt2_type != CODEPOINT_TYPE_LETTER && cpt2_type != CODEPOINT_TYPE_DIGIT && cpt2_type != CODEPOINT_TYPE_UNIDENTIFIED) {
                 pos += (cpt == ' ');
-                while(cpt2_type != CODEPOINT_TYPE_WHITESPACE && cpt2_type != CODEPOINT_TYPE_LETTER && cpt2_type != CODEPOINT_TYPE_DIGIT && cpt2_type != CODEPOINT_TYPE_UNIDENTIFIED)
+                while (cpt2_type != CODEPOINT_TYPE_WHITESPACE && cpt2_type != CODEPOINT_TYPE_LETTER && cpt2_type != CODEPOINT_TYPE_DIGIT && cpt2_type != CODEPOINT_TYPE_UNIDENTIFIED) {
                     cpt2_type = _get_cpt_type(++pos);
+                }
                 char32_t cpt2 = _get_cpt(pos);
-                while(cpt2 == '\r' || cpt2 == '\n')
+                while (cpt2 == '\r' || cpt2 == '\n') {
                     cpt2 = _get_cpt(++pos);
+                }
                 _add_token(pos);
                 continue;
             }
@@ -433,8 +441,9 @@ static std::vector<size_t> unicode_regex_split_custom_llama3(const std::string &
             size_t last_end_r_or_n = 0;
             while (_get_cpt_type(pos+num_whitespaces) == CODEPOINT_TYPE_WHITESPACE) {
                 char32_t cpt2 = _get_cpt(pos+num_whitespaces);
-                if (cpt2 == '\r' || cpt2 == '\n')
+                if (cpt2 == '\r' || cpt2 == '\n') {
                     last_end_r_or_n = pos + num_whitespaces + 1;
+                }
                 num_whitespaces++;
             }
 
@@ -446,14 +455,14 @@ static std::vector<size_t> unicode_regex_split_custom_llama3(const std::string &
             }
 
             // regex: \s+(?!\S)
-            if(num_whitespaces > 1 && _get_cpt(pos+num_whitespaces) != 0) {
+            if (num_whitespaces > 1 && _get_cpt(pos+num_whitespaces) != 0) {
                 pos += num_whitespaces - 1;
                 _add_token(pos);
                 continue;
             }
 
             // regex: \s+
-            if(num_whitespaces > 0) {
+            if (num_whitespaces > 0) {
                 pos += num_whitespaces;
                 _add_token(pos);
                 continue;
```
</details>

File: unicode.cpp
Comment (ggerganov) [reply unicode.cpp:227]: Yup, and `text_collapsed` should be computed lazily (thx @ibob for the hint). But we can do all these optimizations after we merge this. For let's focus on verifying the correctness with some large diverse corpus

**Comment by s-kostyaev** [comment]
@jaime-m-p good solution for bruteforce tests is coverage-guided fuzzing, see https://llvm.org/docs/LibFuzzer.html
Founded problems may later be converted to unit tests.

**Comment by ggerganov** [comment]
@jaime-m-p PTAL at #7036 

> Where do these code point tables come from?

Not sure. I've added a python script to regenerate the ranges:

```bash
python3 scripts/gen-unicode-data.py
```

I've also added a basic script to compare `llama.cpp` and `transformers` tokenization:

```bash
./tests/test-tokenizer-0.sh llama-bpe ./wikitext-2-raw/wiki.train.raw
```

You can obtain `wikitext` dataset using `./scripts/get-wikitext-2.sh`

We don't have to use `wikitext` specifically - if we generate or find a more diverse dataset, we can use that for tokenization tests

**Comment by ggerganov** [comment]
Let's change the target branch to `master`

**Comment by reneleonhardt** [comment]
> @jaime-m-p good solution for bruteforce tests is coverage-guided fuzzing, see https://llvm.org/docs/LibFuzzer.html Founded problems may later be converted to unit tests.

A regex engine is forbidden because parsing by hand and syncing unicode tables is easier, but a huge library with dozens of files is an acceptable `single-header/single-source` solution? 🤔
https://github.com/llvm/llvm-project/tree/main/compiler-rt/lib/fuzzer

**Comment by ggerganov** [comment]
> What do you think about using bitfields?

Yes, seems like a good idea

p.s. I changed the target branch to `master`

**Review by ggerganov — APPROVED** [review]
I discovered fail case using `llama-bpe`: `Cửa Việt`

```
src: 'Cửa Việt'
res: 'Cửa Việt'
tok: 34 91163 11655 26298 83 
main : failed test:    'Cửa Việt'
main : detokenized to: 'Cửa Việt' instead of 'Cửa Việt'
main : expected tokens:     34 'C',  91163 'ửa', 101798 ' Việt', 
main : got tokens:          34 'C',  91163 'ửa',  11655 ' Vi',  26298 'ệ',     83 't', 
```

This also fails on `master`

Other than that, I think this PR is great - nice job!
I couldn't test the random tokenizer tests because I'm on Mac atm, but will give this a try later on my Linux box.

Will merge this after the CI is green

**Comment by mofosyne** [comment]
gg's intent is to merge after the CI is green... all check has passsed... merging

**Pull request merged**
